src/shared/sam/activation.ts

```diff
@@ -27,13 +30,55 @@ import { SamDebugConfigProvider } from './debugger/awsSamDebugger'
 import { addSamDebugConfiguration } from './debugger/commands/addSamDebugConfiguration'
 import { SamDebugSession } from './debugger/samDebugSession'
 import { AWS_SAM_DEBUG_TYPE } from './debugger/awsSamDebugConfiguration'
+import { ext } from '../extensionGlobals'
+
+const YAML_PLUGIN = 'redhat.vscode-yaml'
+const YAML_PROMPT = 'dontShowYamlPromptEver'
 
 /**
  * Activate SAM-related functionality.
  */
 export async function activate(ctx: ExtContext): Promise<void> {
     initializeSamCliContext({ settingsConfiguration: ctx.settings })
 
+    const dontPromptEverAgain = ext.context.globalState.get<boolean>(YAML_PROMPT)
+    // only pop this up in VS Code and Insiders since other VS Code-like IDEs (e.g. Theia) may not have a marketplace or contain the YAML plugin
+    const isVsCode = vscode.env.appName && vscode.env.appName.startsWith('Visual Studio Code') ? true : false
+
+    if (!dontPromptEverAgain && !vscode.extensions.getExtension(YAML_PLUGIN) && isVsCode) {
+        // these will all be disposed immediately after showing one so the user isn't prompted more than once per session
+        const yamlPromptDisposables: vscode.Disposable[] = []
+
+        vscode.workspace.onDidOpenTextDocument(
+            async (doc: vscode.TextDocument) => {
+                if (doc.fileName.endsWith('template.yaml') || doc.fileName.endsWith('template.yml')) {
+                    promptInstallYamlPlugin()
+                    for (const prompt of yamlPromptDisposables) {
+                        prompt.dispose()
+                    }
+                }
+            },
+            undefined,
+            yamlPromptDisposables
+        )
+
+        vscode.window.onDidChangeActiveTextEditor(
+            async (editor: vscode.TextEditor | undefined) => {
+                if (editor) {
+                    const fileName = editor.document.fileName
+                    if (fileName.endsWith('template.yaml') || fileName.endsWith('template.yml')) {
+                        promptInstallYamlPlugin()
+                        for (const prompt of yamlPromptDisposables) {
+                            prompt.dispose()
+                        }
+                    }
```

Comment:
could DRY this a bit by putting the shared logic from line 53 in a local function.